DIPs/1NNN-RAC.md

```diff
@@ -0,0 +1,333 @@
+
+When not active it places it into DllImport mode as if ``extern`` was also annotated.
+
+To standardise the identifiers in use, three new prefixes are added to the specification, each will have instances for libc, druntime and Phobos that will be provided by the compiler automatically.
+
+The prefixes are ``Have_``, ``InBinary_``, and ``Compiling_``. The suffixes for these three will be a logical package.
+
+- ``Have_`` is used when a given logical package is available as a dependency during linking.
+- ``InBinary_`` is used when a given logical package has its symbols in the currently compiling binary.
+- ``Compiling_`` is used when a given logical package is currently being compiled.
+
+Between these three versions it is possible to cover any corner case where a symbol must be viewed internally instead of out of binary.
+
+The D interface generator will need to support the insertion of the version argument ``InBinary_``. This will be needed to accurately represent C files by a D module. The mechanism to specify what the suffix is, is not determined here. It may depend upon C preprocessor capability.
+
+#### Positive Notation
+
+To assign a symbol that it is exported, use the ``export`` attribute to annotate a D symbol as DllExport.
+
+The export annotation must not be a visibility modifier.
+If it is a visibility modifier you are required to expose internal implementation details that could result in unsafe operation of a codebase by external parties at the language level.
+
+For a given encapsulation unit (struct, class, union, module), if any members are marked export, then all generated symbols (``TypeInfo``, ``__initZ``, ``opCmp`` ext.) but not ``ModuleInfo`` must also be exported. If generated symbols are not exported when an associated symbol is exported, it will result in linker errors that have no possible solution without the use of a linker script.
+
+By default all symbols are hidden, if you need to explicitly annotate hidden use the UDA provided in ``core.attributes``.
+Without this attribute you would have to annotate export on _every_ symbol that you want exported instead of at the scope and then disallow only the ones that are not desired.
+
+#### Negative Notation
+
+The visibility override switch is in use, set to export everything.
+
+To annotate a symbol as hidden use the UDA provided in ``core.attributes``.
+
+Until druntime and Phobos can be fully annotated with export appropriately and tested, this approach will be in use for these libraries.
+
+### Inlining
+
+Consider an out of binary module:
+
+```d
+pragma(inline, true)
+export extern void inlineable() {
+  noInline;
+}
+
+@hidden void noInline();
+```
+
+In this case, ``noInline`` function is not accessible to ``inlineable`` if it was inlined across binary lines. The result of this would be a linker error.
+
+The compiler must not inline a function from an out of binary module if it refers to non-exported symbols.
```

Comment:
And what happens if a user writes `pragma(inline, true)` anyway?

Reply:
In practice, that pragma is treated as a hint to the compiler.
See Johan's comment about it: https://forum.dlang.org/post/szhjqciicrupbmqwpkeq@forum.dlang.org

The language might be stronger than needed, and probably shouldn't end up in the spec.

It'll simply be ignored like any optimization hint should be when it will do the wrong thing.